design/nfs/nfs-provisioner-controlled-by-operator.md

```diff
@@ -0,0 +1,115 @@
+│       └── data (from PV-B)
+└── another-export
+```
+
+Since those directories are not in the NFS Provisioner container but in NFS Server container, NFS Provisioner cannot directly create sub-directory for that. The solution is by mounting the whole of underlying NFS share/export directory so that the NFS Provisioner can create sub-directory for each provisioned PVs.
+
+### Original Issue
+
+- https://github.com/rook/rook/issues/4982
+
+### Goals
+
+- Make NFS Provisioner to create sub-directory for each provisioned PVs.
+- Make NFS Provisioner to use sub-directory for each provisioned PVs instead of use underlying directory.
+- Improve reliability of NFS Provisioner.
+
+### Non-Goals
+
+- NFS Operator manipulates uncontrolled resources.
+
+## Proposal details
+
+The approach will be similar to [Kubernetes NFS Client Provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/nfs-client) which the provisioner mounts the whole of NFS share/export into the provisioner pod (by kubelet) then the provisioner can create the appropriate sub-directory for each provisioned PVs. Currently Rook NFS Provisioner deployed independently and before NFS Server itself we cannot mount the NFS share because we don't know what NFS Server IP and the share/export directory. 
+
+The idea is to make NFS Provisioner controlled by the operator. So when NFS Server is created, the operator also creates its provisioner which mount each NFS share/export. And the rest NFS Provisioner can create sub-directory for provisioned PV.
+
+This is the example NFS Server
+
+```yaml
+apiVersion: nfs.rook.io/v1alpha1
+kind: NFSServer
+metadata:
+  name: rook-nfs
+  namespace: rook-nfs
+spec:
+  replicas: 1
+  exports:
+  - name: share1
+    ...
+    persistentVolumeClaim:
+      claimName: nfs-default-claim
+  - name: share2
+    ...
+    persistentVolumeClaim:
+      claimName: nfs-another-claim
+```
+
+And the operator will creates the provisioner deployment like
+
+```yaml
+kind: Deployment
+apiVersion: apps/v1
+metadata:
+  name: rook-nfs-provisioner
+  namespace: rook-nfs
+spec:
+  ...
+    spec:
+      ....
+      containers:
+      - name: rook-nfs-provisioner
+        image: rook/nfs:master
+        args: ["nfs", "provisioner","--provisioner=nfs.rook.io/nfs-server-provisioner"]
+      volumes:
+        - name: share1
+          nfs:
+            server: <NFS_SERVER_IP>
+            path: /export/nfs-default-claim
+        - name: share2
+          nfs:
+            server: <NFS_SERVER_IP>
+            path: /export/nfs-another-claim
+```
+
+The provisioner deployment will be created on the same namespace with the NFS server and possibility has a same privileges. Since provisioner automatically created by operator the provisioner deployment name and provisioner name flag (`--provisioner`) value will depend on NFSServer name wich deployment name will be added suffix `-provisioner` and provisioner name will started by `nfs.rook.io/` and ended by `-provisioner`
```

Comment:
```suggestion
The provisioner deployment will be created in the same namespace as the NFS server and with the same privileges. Since the provisioner is automatically created by the operator, the provisioner deployment name and provisioner name flag (`--provisioner`) value will depend on NFSServer name. The provisioner deployment name will have an added suffix of `-provisioner` and the provisioner name will start with `nfs.rook.io/`.
```